scripts/apps/monitoring/directives/WidgetGroup.js

```diff
@@ -0,0 +1,364 @@
+    $location, $anchorScroll, activityService, $rootScope, gettextCatalog, datetime, metadata) {
+    const services = {
+        search: search,
+        api: api,
+        superdesk: superdesk,
+        desks: desks,
+        cards: cards,
+        $timeout: $timeout,
+        $q: $q,
+        $location: $location,
+        $anchorScroll: $anchorScroll,
+        activityService: activityService,
+        $rootScope: $rootScope,
+        gettextCatalog: gettextCatalog,
+        datetime: datetime,
+        metadata: metadata
+    };
+
+    return {
+        scope: {
+            stage: '=',
+            total: '=',
+            allowed: '=',
+            showEmpty: '=?',
+            maxItems: '=?',
+            selected: '=?',
+            action: '&',
+            filter: '='
+        },
+        link: function(scope, elem) {
+            var criteria;
+
+            scope.page = 1;
+            scope.fetching = false;
+            scope.items = [];
+
+            /**
+              * Generates Identifier to be used by track by expression.
+              */
+            scope.generateTrackByIdentifier = function(item) {
+                return search.generateTrackByIdentifier(item);
+            };
+
+            scope.preview = function(item) {
+                superdesk.intent('preview', 'item', item);
+            };
+
+            scope.edit = function(item) {
+                if (item._type === 'ingest') {
+                    var activity = superdesk.findActivities({action: 'list', type: 'ingest'}, item)[0];
+
+                    activityService.start(activity, {data: {item: item}}).then((item) => {
+                        initEdit(item);
+                    });
+                } else {
+                    initEdit(item);
+                }
+
+                function initEdit(item) {
+                    superdesk.intent('edit', 'item', item).then(null, () => {
+                        superdesk.intent('view', 'item', item);
+                    });
+                }
+            };
+
+            function getProvider(criteria) {
+                var provider = 'archive';
+
+                if (scope.stage.type && (desks.isOutputType(scope.stage.type) || scope.stage.type === 'search')) {
+                    provider = 'search';
+                }
+
+                if (criteria.repo && criteria.repo.indexOf(',') === -1) {
+                    provider = criteria.repo;
+                }
+
+                return provider;
+            }
+
+            function queryItems(queryString) {
+                var oldSize;
+
+                criteria = cards.criteria(scope.stage, queryString);
+                if (!scope.fetching) {
+                    scope.setLoading(true);
+                }
+
+                oldSize = criteria.source.size;
+                criteria.source.from = 0;
+
+                if (scope.page > 0 && criteria.source) {
+                    criteria.source.size = (scope.page - 1) * oldSize;
```

Comment:
I think there would be a difference in server returning 100 or 1000 rows. Payload is increasing with that many rows is one thing that makes elastic server and network busy. Also when the stories are fetched we call `enhance_items` functions to attach additional information, this increases CPU utilization. If few users are scrolling then it would affect overall performance.

I agree with Petr the we should try to get 25-50 items everytime but on client we can keep a larger number of items.